File: PHP/Inkassogram/Bookkeeping/BookkeepingClient.php
<?php
/**
 * Inkassogram Bookkeeping API Client
 *
 * @author      Johan Sall Larsson <johan@inkassogram.se>
 * @author      Simon Stal <simon@inkassogram.se>
 * @copyright   2017 Inkassogram AB (publ)
 * @version     2.0.0
 * @package     Inkassogram
 *
 * MIT LICENSE
 *
 * Permission is hereby granted, free of charge, to any person obtaining
 * a copy of this software and associated documentation files (the
 * "Software"), to deal in the Software without restriction, including
 * without limitation the rights to use, copy, modify, merge, publish,
 * distribute, sublicense, and/or sell copies of the Software, and to
 * permit persons to whom the Software is furnished to do so, subject to
 * the following conditions:
 *
 * The above copyright notice and this permission notice shall be
 * included in all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
 * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
 * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
 * NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE
 * LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
 * OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION
 * WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */
 
namespace Inkassogram\Bookkeeping; 

/**
 * The Inkassogram Invoice Bookkeeping API client.
 *
 * @package Inkassogram
 * @author  Johan Sall Larsson <johan@inkassogram.se>
 * @author  Simon Stal <simon@inkassogram.se>
 * @since   2.0.0
 */
class BookkeepingClient
{

    /**
     * The customer number.
     * @var int
     */
    private $customerNumber;
    
    /**
     * The authorization token.
     * @var string
     */
    private $authToken;
    
    /**
     * The IP address.
     * @var string
     */
    private $ipAddress;
    
    /**
     * If the XML data should be validated
     * by the XSD file before the request.
     * @var bool
     */
    private $validateXsdBeforeRequest;
    
    /**
     * The XML data that is sent in the POST request.
     * @var string
     */
    private $lastRequest;
    
    /********************************************************************************
     * Public Functions
     *******************************************************************************/
    
    /**
     * Creates an invoice.
     * @param CreateInvoiceRequest $request
     */
    public function CreateInvoice(Contract\CreateInvoiceRequest $request)
    {
        return $this->sendRequest('https://api.inkassogram.se/API/createInvoiceBookkeeping', $request);
        //return $this->sendRequest('http://10.99.196.64/inkassogram.se/www.inkassogram.se/API/createInvoiceBookkeeping', $request);
    }
    
    /**
     * Creates an invoice.
     * @param CreditInvoiceRequest $request
     */
    public function CreditInvoice(Contract\CreditInvoiceRequest $request)
    {
        return $this->sendRequest('https://api.inkassogram.se/API/creditInvoice', $request);
    }
     
    /********************************************************************************
     * Getters and setters
     *******************************************************************************/
     
    /**
     * Returns the customer number.
     * @return int
     */
    public function getCustomerNumber()
    {
        return $this->customerNumber;
    }
    
    /**
     * Sets the customer number.
     * @return this
     */
    public function setCustomerNumber($customerNumber)
    {
        $this->customerNumber = $customerNumber;
        return $this;
    }
    
    /**
     * Returns the authorization token.
     * @return string
     */
    public function getAuthToken()
    {
        return $this->authToken;
    }
    
    /**
     * Sets the authorization token.
     * @return this
     */
    public function setAuthToken($authToken)
    {
        $this->authToken = $authToken;
        return $this;
    }
    
    /**
     * Returns the IP address.
     * @return string
     */
    public function getIpAddress()
    {
        return $this->ipAddress;
    }
    
    /**
     * Sets the IP address.
     * @param this
     */
    public function setIpAddress($ipAddress)
    {
        $this->ipAddress = $ipAddress;
        return $this;
    }
    
    /**
     * Returns the validate XML data before request.
     * @return bool
     */
    public function getValidateXsdBeforeRequest()
    {
        return $this->validateXsdBeforeRequest;
    }
    
    /**
     * Sets the validate XML data before request.
     * @param  bool $validateXsdBeforeRequest
     * @return this
     */
    public function setValidateXsdBeforeRequest($validateXsdBeforeRequest)
    {
        $this->validateXsdBeforeRequest = $validateXsdBeforeRequest;
        return $this;
    }
    
    /**
     * Returns the XML data in the POST request.
     * @return string
     */
    public function getLastRequest()
    {
        return $this->lastRequest;
    }
    
    /**
     * Sets the XML data in the POST request.
     * @param string $lastRequest
     */
    private function setLastRequest($lastRequest)
    {
        $this->lastRequest = $lastRequest;
    }
    
    /********************************************************************************
     * Private Functions
     *******************************************************************************/
     
    /**
     * Sends the POST request.
     * @param    string                $serviceUrl
     * @param    object                $obj
     * @return   SimpleXMLElement
     */
    private function sendRequest($serviceUrl, $obj)
    {
        $serializer = new \Inkassogram\Bookkeeping\Serializer\XmlSerializer();
        $xmlDoc = $serializer->serialize($obj);
        $post = $xmlDoc->asXML();
        $this->setLastRequest($post);        
        if ($this->getValidateXsdBeforeRequest()) {
            $this->validateXsd($xmlDoc, $post);
        }
        $ch = curl_init();
        curl_setopt_array($ch, array( 
            CURLOPT_POST => 1, 
            CURLOPT_HEADER => 0, 
            CURLOPT_URL => $serviceUrl, 
            CURLOPT_FRESH_CONNECT => 1, 
            CURLOPT_RETURNTRANSFER => 1, 
            CURLOPT_FORBID_REUSE => 1, 
            CURLOPT_TIMEOUT => 9, 
            CURLOPT_POSTFIELDS => $post,
            CURLOPT_HTTPHEADER => array(
                'Content-Type: text/xml', 
                sprintf('customerNo: %s', $this->getCustomerNumber()), 
                sprintf('key: %s', md5($this->getIpAddress() . date('Ymd') . $this->getAuthToken())), 
                sprintf('Content-length: %s', strlen($post))
            )
        ));
        if (! $result = curl_exec($ch)) { 
            trigger_error(curl_error($ch)); 
        } 
        curl_close($ch); 
        $retval = simplexml_load_string($result);
        return $retval;
    }
    
    /**
     * Validates the XML by XSD schema.
     * @param    SimpleXMLElement        $xmlDoc
     * @param    string                  $data
     * @throws   XsdValidationException
     */
    private function validateXsd($xmlDoc, $data)
    {
        $attr = $xmlDoc->attributes();
        $schemaLocation = explode(' ', $attr["schemaLocation"]);
        $doc = new \DomDocument;
        $doc->loadXML($data);
        libxml_use_internal_errors(true);
        if (!$doc->schemaValidate($schemaLocation[1])) {
            return true;
        } else {
            $errors = libxml_get_errors();
            throw new \Inkassogram\Bookkeeping\Exception\XsdValidationException("XSD validation failed!\n\n" . var_export(error_get_last(), true));
        }
    }
    
}

File: Java/src/main/java/se/inkassogram/api/client/BookkeepingClient.java
package se.inkassogram.api.client;

/**
 * Inkassogram Bookkeeping API Client
 *
 * @author    <a href="mailto:dev@inkassogram.se">Inkassogram</a>
 * @version   1.0.0
 *
 * MIT LICENSE
 *
 * Permission is hereby granted, free of charge, to any person obtaining
 * a copy of this software and associated documentation files (the
 * "Software"), to deal in the Software without restriction, including
 * without limitation the rights to use, copy, modify, merge, publish,
 * distribute, sublicense, and/or sell copies of the Software, and to
 * permit persons to whom the Software is furnished to do so, subject to
 * the following conditions:
 *
 * The above copyright notice and this permission notice shall be
 * included in all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
 * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
 * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
 * NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE
 * LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
 * OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION
 * WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

import java.io.IOException;
import java.io.InputStream;
import java.text.SimpleDateFormat;
import java.util.Date;
import java.util.List;
import java.util.Properties;

import javax.xml.bind.JAXBElement;
import javax.xml.bind.JAXBException;

import org.apache.http.HttpEntity;
import org.apache.http.HttpResponse;
import org.apache.http.client.ClientProtocolException;
import org.apache.http.client.methods.HttpPost;
import org.apache.http.entity.StringEntity;
import org.apache.http.impl.client.DefaultHttpClient;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import se.inkassogram.api.createinvoicebookkeeping.MethodCallType;
import se.inkassogram.api.utils.EncryptionUtils;
import se.inkassogram.api.utils.XmlUtils;

/**
 * The Bookkeeping API client. 
 *
 * @author <a href="mailto:dev@inkassogram.se">Inkassogram</a>
 * @since  1.0.0
 */
public class BookkeepingClient {
    
    /**
     * Logging.
     */
    static Logger log = LoggerFactory.getLogger(BookkeepingClient.class);
    
    /**
     * The customer number.
     */
    Integer customerNo;
    
    /**
     * The Authentication token/key.
     */
    String authToken;
    
    /**
     * The IP address.
     */
    String ipAddress;
    
    /**
     * The .properties file name.
     */
    static final String defaultPropertiesFile = "bookkeeping-api.properties";
    
    /**
     * The read properties.
     */
    static Properties properties;
    
    /**
     * Read lock to prevent re-initialization.
     */
    static final Object readLock = new Object();
    
    /**
     * The endpoint URL.
     */
    static final String createInvoiceEndpointUrl = "https://api.inkassogram.se/API/createInvoiceBookkeeping";

    /**
     * The endpoint URL.
     */
    static final String creditInvoiceEndpointUrl = "https://api.inkassogram.se/API/creditInvoice";
    
    /**
     * Default constructor.
     */
    public BookkeepingClient() 
    {
        
    }
    
    /**
     * Contructor with params.
     * 
     * @param customerNo
     * @param authToken
     * @param ipAddress
     */
    public BookkeepingClient(Integer customerNo, String authToken, String ipAddress)
    {
        this.setCustomerNo(customerNo);
        this.setAuthToken(authToken);
        this.setIpAddress(ipAddress);
    }
    
    /**
     * Creates an invoice.
     * 
     * @param  request
     * @return MethodCallType
     * @throws ClientProtocolException
     * @throws IOException
     */
    public se.inkassogram.api.createinvoicebookkeeping.MethodCallType createInvoice(se.inkassogram.api.createinvoicebookkeeping.RequestType request) 
            throws ClientProtocolException, IOException 
    {
        se.inkassogram.api.createinvoicebookkeeping.ObjectFactory objectFactory = new se.inkassogram.api.createinvoicebookkeeping.ObjectFactory();
        se.inkassogram.api.createinvoicebookkeeping.MethodCallType methodCallType = new se.inkassogram.api.createinvoicebookkeeping.MethodCallType();
        methodCallType.setMethodName("createInvoice");
        methodCallType.setRequest(request);        
        JAXBElement<se.inkassogram.api.createinvoicebookkeeping.MethodCallType> methodCall = objectFactory.createMethodCall(methodCallType);
        MethodCallType callType =  sendRequest(createInvoiceEndpointUrl, methodCall, se.inkassogram.api.createinvoicebookkeeping.MethodCallType.class);
        return callType;
    }
    
    /**
     * Credits an invoice.
     * 
     * @param  request
     * @return List<ResponseType>
     * @return MethodCallType
     * @throws ClientProtocolException
     * @throws IOException
     */
    public se.inkassogram.api.creditinvoice.MethodCallType creditInvoice(List<se.inkassogram.api.creditinvoice.RequestType> request) 
            throws ClientProtocolException, IOException
    {
        se.inkassogram.api.creditinvoice.ObjectFactory objectFactory = new se.inkassogram.api.creditinvoice.ObjectFactory();
        se.inkassogram.api.creditinvoice.MethodCallType methodCallType = new se.inkassogram.api.creditinvoice.MethodCallType();
        methodCallType.setMethodName("creditInvoice");
        methodCallType.getRequest().addAll(request);
        JAXBElement<se.inkassogram.api.creditinvoice.MethodCallType> methodCall = objectFactory.createMethodCall(methodCallType);
        return sendRequest(creditInvoiceEndpointUrl, methodCall, se.inkassogram.api.creditinvoice.MethodCallType.class);
    }
    
    /**
     * Sends a POST request.
     * 
     * @param  object
     * @throws ClientProtocolException 
     * @throws IOException 
     */
    <TRequest, TResponse> TResponse sendRequest(String url, JAXBElement<TRequest> requestObject, Class<TResponse> responseObject) 
            throws IOException, ClientProtocolException 
    {
        loadProperties();
        DefaultHttpClient httpClient = new DefaultHttpClient();
        HttpPost httpPost = new HttpPost(url);
        SimpleDateFormat formatter = new SimpleDateFormat("yyyyMMdd");
        
        String key = EncryptionUtils.hash((getIpAddress() + formatter.format(new Date()) + getAuthToken()));
        httpPost.addHeader("accept", "application/xml");
        httpPost.addHeader("customerNo", getCustomerNo().toString());
        httpPost.addHeader("key", key);
        
        try {
            String data = XmlUtils.serialize(requestObject);
            StringEntity stringEntity = new StringEntity(data);
            stringEntity.setContentType("application/xml");
            stringEntity.setContentEncoding("UTF-8");
            httpPost.setEntity(stringEntity);
            log.debug(String.format("XML request: ", data));
            HttpResponse response = httpClient.execute(httpPost);
            HttpEntity entity = response.getEntity();
            if (entity != null) {
                InputStream instream = entity.getContent();
                try {
                    InputStream is = response.getEntity().getContent();
                    return XmlUtils.deserialize(is, responseObject);
                } catch (IOException e) {
                    throw e;
                } catch (RuntimeException e) {
                    httpPost.abort();
                    throw e;
                } finally {
                    try { 
                        instream.close();  
                    }
                    catch (Exception e) {
                        log.warn("Cannot close response stream.", e);    
                    }
                }
            } else {
                throw new ClientProtocolException("Response contains no content");
            }
        } catch (JAXBException e) {
            throw new ClientProtocolException("Malformed XML", e);
        } finally {
            httpClient.getConnectionManager().shutdown();
        }
    }

    /**
     * Loads the .properties file and sets authToken, customerNo
     * and ipAddress.
     * 
     * @throws IOException 
     */
    void loadProperties() throws IOException {
        if (
            getCustomerNo() == null ||
            getAuthToken() == null || 
            getIpAddress() == null
        ) {
            if (properties == null) {
                synchronized(readLock) {
                    if (properties == null) {
                        log.debug(String.format("Loading properties from %s.", defaultPropertiesFile));
                        ClassLoader loader = BookkeepingClient.class.getClassLoader();
                        InputStream in = loader.getResourceAsStream(defaultPropertiesFile);
                        properties = new Properties();
                        properties.load(in);
                        this.setCustomerNo(Integer.parseInt(properties.getProperty("customerNo")));
                        this.setAuthToken(properties.getProperty("authToken"));
                        this.setIpAddress(properties.getProperty("ipAddress"));    

                        System.out.println(properties.getProperty("authToken"));
System.out.println(properties.getProperty("ipAddress"));
System.out.println(properties.getProperty("customerNo"));

                }
            }
        }
    }
    }

    /**
     * Returns the customer number.
     */
    public Integer getCustomerNo() {
        return customerNo;
    }

    /**
     * Sets the customer number.
     * @param customerNo
     */
    public void setCustomerNo(Integer customerNo) {
        this.customerNo = customerNo;
    }

    /**
     * Returns the authentication token/key.
     */
    public String getAuthToken() {
        return authToken;
    }

    /**
     * Sets the authentication token/key.
     * @param authToken
     */
    public void setAuthToken(String authToken) {
        this.authToken = authToken;
    }

    /**
     * Returns the IP address.
     */
    public String getIpAddress() {
        return ipAddress;
    }

    /**
     * Sets the IP address.
     * @param ipAddress
     */
    public void setIpAddress(String ipAddress) {
        this.ipAddress = ipAddress;
    }
    
}
